src/extras/named-register.js

```diff
@@ -16,15 +16,22 @@
   SystemJS.prototype = systemJSPrototype;
   System = new SystemJS();
 
+  let isInstantiating = false;
+
   const register = systemJSPrototype.register;
   systemJSPrototype.register = function (name, deps, declare) {
     if (typeof name !== 'string')
       return register.apply(this, arguments);
 
-    this.registerRegistry[name] = [deps, declare];
```

Comment:
So the approach would be something like having a `let lastNamedDefine` at the top.

Then in the `register` function here it would be -

```js
this.registerRegistry[name] = lastNamedDefine = [deps, declare];
```

Then we will need a hook to `getRegister` here (https://github.com/systemjs/systemjs/blob/master/docs/hooks.md#getregister---deps-string-declare-function), which would provide the `lastNamedDefine` if the `getRegister` function would have returned nothing.

Then we also need to very carefully reset `lastNamedDefine` to `undefined` in the `getRegister` method itself (which is a sort of sync version of how you were trying to use instantiate end specially to avoid the timing issues).